src/nemo-connectivity/mobiledataconnection.cpp

```diff
@@ -159,6 +160,53 @@ void MobileDataConnectionPrivate::updateServiceProviderName()
     }
 }
 
+void MobileDataConnectionPrivate::updateTechnology()
+{
+    NetworkTechnology *newTech;
+    QString technologyPath = QString("cellular");
+    bool oldPowered = false;
+    bool initializing = true;
+
+    newTech = networkManager.getTechnology(technologyPath);
+
+    qCDebug(CONNECTIVITY, "####### update technology from %p to %p", networkTechnology, newTech);
+
+    if (networkTechnology == newTech)
+        return;
+
+    if (networkTechnology) {
+        /* ConnMan restart has happened and in lib networkTechnology is gone */
+        if (newTech) {
+            oldPowered = networkTechnology->powered();
+            QObject::disconnect(networkTechnology, &NetworkTechnology::poweredChanged, q, 0);
+        }
+
+        initializing = false;
+    }
+
+    networkTechnology = newTech;
+
+    if (networkTechnology) {
+        QObject::connect(networkTechnology, &NetworkTechnology::poweredChanged, q, [=](bool powered) {
+            techPoweredChanged(powered);
+        });
+
+        bool powered = networkTechnology->powered();
+
+        /* Do powered state change only if changing tech with different state */
+        if (powered != oldPowered && !initializing) {
```

Comment:
I feel that we should call techPoweredChanged() always when going from NULL networkTechnology to non-NULL.

Reply:
I need to re-check this but if we do it always then the state cycling happens since it would be called before ofono has retrieved the state. I think it is better to rely on the changes made by ofono than to set connman into state with information that is not certain yet?

Reply:
Erm, yeah, I missed that the initializing is a local var, probably nevermind then. If connman signals us once we have some actual data, then all should be fine.

Reply:
And in general, I really think that the wrong powered states should be handled elsewhere than in Settings app :)